orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageContext.java

```diff
@@ -56,7 +57,10 @@ public Object get(@Nullable Object key) {
     if (delegate().containsKey(key)) {
       return super.get(key);
     } else {
-      return stage.ancestors().stream()
+      if (ancestors == null) {
```

Comment:
That's true. I think caching in the stage is just a fraught since the stage wouldn't know when a new stage is inserted into an execution upstream which could happen during stage planning.

The interesting bit is that, technically (today), no mutation will alter the behavior of this `get`ter, the reason is that all this "lookup" in prior stage "outpus" is really only used in bake and evalvars stages today, and it's not like an upstream bake stage will complete while a stage's task is being evaluated (which is the lifetime of stagecontext).

This is all WAY too convoluted. But in the end, I agree with you that even though (I believe) this is safe today, it might not be tomorrow and the code has a considerable "surprise" factor. So I will take it out for now and if I think of a better way of doing this will put up another PR (the perf improvement from recursion removal is pretty considerable already)

Reply:
Oh, i have an idea... @ezimanyi what do you think about caching the total number of stages in the execution and invalidating the cache when that changed? That actually seems safe. It's not immune to stages updating their `requisiteStageRefIds` but that doesn't happen AFAIK

Reply:
I think that sounds reasonable...

I was just looking that how stages get added to an `Execution` and if I'm understanding correctly, there's no `addStage` directly, but rather callers use `getStages` and mutate the resulting array.  Based on that, we really don't have a lot of control over how/when to invalidate the cache (from anywhere...either from stage or from `Execution`).  So I still think this would be a bit risky to cache the value...but given that the current usage pattern appears to be to only ever add stages to the array, caching based on number of stages is probably a reasonable solution if the performance benefits warrant it.